### PR TITLE
meta: add an in-memory autoID allocator

### PR DESCRIPTION
**Pull request by djshow832**
### What problem does this PR solve?

Problem Summary: add an in-memory autoID allocator for temporary tables.
Part of https://github.com/pingcap/tidb/issues/24169

### What is changed and how it works?

What's Changed: define an `inMemoryAllocator`.

How it Works: allocate autoID in memory.

### Related changes

- N/A

### Check List 

Tests 

- Unit test

Side effects

- N/A

### Release note 

- Add an in-memory autoID allocator.

### CODE REVIEW

File: meta/autoid/autoid.go [hunk 1]
```diff
@@ -912,6 +912,133 @@ func (alloc *allocator) alloc4Sequence(tableID int64) (min int64, max int64, rou
 	return newBase, newEnd, round, nil
 }
 
+// NewAllocatorFromTempTblInfo creates an in-memory allocator from a temporary table info.
+func NewAllocatorFromTempTblInfo(tblInfo *model.TableInfo) Allocator {
+	hasRowID := !tblInfo.PKIsHandle && !tblInfo.IsCommonHandle
+	hasAutoIncID := tblInfo.GetAutoIncrementColInfo() != nil
+	// Temporary tables don't support auto_random and sequence.
+	if hasRowID || hasAutoIncID {
+		return &inMemoryAllocator{
```
Comment (AilinKid) [inline meta/autoid/autoid.go:921]: the rowid and autoid should be separated? @tangenta

File: meta/autoid/autoid.go [hunk 2]
```diff
@@ -912,6 +912,133 @@ func (alloc *allocator) alloc4Sequence(tableID int64) (min int64, max int64, rou
...
+
+	// CalcNeededBatchSize calculates the total batch size needed.
+	n1 := CalcNeededBatchSize(alloc.base, int64(n), increment, offset, alloc.isUnsigned)
+
+	// Condition alloc.base+n1 > alloc.end will overflow when alloc.base + n1 > MaxInt64. So need this.
+	if math.MaxUint64-uint64(alloc.base) <= uint64(n1) {
+		return 0, 0, ErrAutoincReadFailed
+	}
+
+	logutil.Logger(context.TODO()).Error("b", zap.Int64("n1", int64(n1)), zap.Int64("base", alloc.base))
```
Comment (AilinKid) [inline meta/autoid/autoid.go:1026]: why so many logger here?

**Comment by AilinKid** [comment]
/home/jenkins/agent/workspace/tidb_ghpr_check/go/src/github.com/pingcap/tidb/meta/autoid/autoid.go:1026:65: unnecessary conversion

File: meta/autoid/autoid.go (hunk 1 above)
Comment (tangenta) [reply meta/autoid/autoid.go:921]: It's fine because they are also not separated in non-temporary tables.

File: meta/autoid/autoid.go [hunk 3]
```diff
@@ -912,6 +912,133 @@ func (alloc *allocator) alloc4Sequence(tableID int64) (min int64, max int64, rou
...
+		return 0, 0, ErrAutoincReadFailed
+	}
+
+	min := alloc.base
+	alloc.base += n1
+	return min, alloc.base, nil
+}
+
+func (alloc *inMemoryAllocator) alloc4Unsigned(n uint64, increment, offset int64) (int64, int64, error) {
+	logutil.Logger(context.TODO()).Error("a", zap.Int64("n", int64(n)), zap.Int64("inc", increment), zap.Int64("off", offset), zap.Int64("base", alloc.base))
```
Comment (tangenta) [inline meta/autoid/autoid.go:1012]: Is this used for debugging?

File: meta/autoid/autoid.go (hunk 2 above)
Comment (djshow832) [reply meta/autoid/autoid.go:1026]: Oops

File: meta/autoid/autoid.go [hunk 4]
```diff
@@ -912,6 +912,133 @@ func (alloc *allocator) alloc4Sequence(tableID int64) (min int64, max int64, rou
 	return newBase, newEnd, round, nil
 }
 
+// NewAllocatorFromTempTblInfo creates an in-memory allocator from a temporary table info.
```
[the remark this answers is not in the document]
Comment (djshow832) [reply meta/autoid/autoid.go]: Addressed.

**Review by tangenta — APPROVED** [review]

**Pull request merged**